cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll

```diff
@@ -219,7 +219,7 @@ abstract class PostUpdateNode extends InstructionNode {
 abstract private class PartialDefinitionNode extends PostUpdateNode, TInstructionNode {
   final Instruction getInstructionOrChi() {
     exists(ChiInstruction chi |
-      // TODO: This should be a non-conflated ChiInstruction once #3123 is merged
+      not chi.isResultConflated() and
```

Comment:
With this change we lose flow in the following case:

```cpp
void sink(int *);

struct B
{
  int *i;
};

void f7(B* b)
{
  b->i = source();
}
void f8()
{
  B* b = new B;
  f7(b);
  sink(b->i); // flow [not detected]
}
```

Because the `BufferMayWriteSideEffect` for `b` in `f7(b)` is melded into all aliased memory:

```
#   15|     r15_1(glval<unknown>) = FunctionAddress[f7]           :
#   15|     r15_2(glval<B *>)     = VariableAddress[b]            :
#   15|     r15_3(B *)            = Load                          : &:r15_2, m14_8
#   15|     v15_4(void)           = Call                          : func:r15_1, 0:r15_3
#   15|     m15_5(unknown)        = ^CallSideEffect               : ~m14_6
#   15|     m15_6(unknown)        = Chi                           : total:m14_6, partial:m15_5
#   15|     v15_7(void)           = ^BufferReadSideEffect[0]      : &:r15_3, ~m15_6
#   15|     m15_8(unknown)        = ^BufferMayWriteSideEffect[0]  : &:r15_3
#   15|     m15_9(unknown)        = Chi                           : total:m15_6, partial:m15_8
#   16|     r16_1(glval<unknown>) = FunctionAddress[sink]         :
#   16|     r16_2(glval<B *>)     = VariableAddress[b]            :
#   16|     r16_3(B *)            = Load                          : &:r16_2, m14_8
#   16|     r16_4(glval<int *>)   = FieldAddress[i]               : r16_3
#   16|     r16_5(int *)          = Load                          : &:r16_4, ~m15_9
#   16|     v16_6(void)           = Call                          : func:r16_1, 0:r16_5
#   16|     m16_7(unknown)        = ^CallSideEffect               : ~m15_9
```

I'm not totally sure how to recover this flow without accepting flow through `Chi` instructions that update all aliased memory.

Interestingly, the flow is reported in the following modified program:

```cpp
void sink(int *);

struct B
{
  int *i;
};

void f7(B* b)
{
  b->i = source();
}
void f8(B* b)
{
  f7(b);
  sink(b->i); // flow
}
```

since the `Chi` following the `BufferMayWriteSideEffect` does not update all aliased memory:

```
#   15|     r15_1(glval<unknown>) = FunctionAddress[f7]          :
#   15|     r15_2(glval<B *>)     = VariableAddress[b]           :
#   15|     r15_3(B *)            = Load                         : &:r15_2, m12_7
#   15|     v15_4(void)           = Call                         : func:r15_1, 0:r15_3
#   15|     m15_5(unknown)        = ^CallSideEffect              : ~m12_4
#   15|     m15_6(unknown)        = Chi                          : total:m12_4, partial:m15_5
#   15|     v15_7(void)           = ^BufferReadSideEffect[0]     : &:r15_3, ~m12_9
#   15|     m15_8(unknown)        = ^BufferMayWriteSideEffect[0] : &:r15_3
#   15|     m15_9(int *)          = Chi                          : total:m12_9, partial:m15_8
#   16|     r16_1(glval<unknown>) = FunctionAddress[sink]        :
#   16|     r16_2(glval<B *>)     = VariableAddress[b]           :
#   16|     r16_3(B *)            = Load                         : &:r16_2, m12_7
#   16|     r16_4(glval<int *>)   = FieldAddress[i]              : r16_3
#   16|     r16_5(int *)          = Load                         : &:r16_4, ~m15_9
#   16|     v16_6(void)           = Call                         : func:r16_1, 0:r16_5
#   16|     m16_7(unknown)        = ^CallSideEffect              : ~m15_6
```

Reply:
Is that a flaw in our support for allocations (https://github.com/Semmle/ql/pull/2797)? I would think that `new B` and a `B *` parameter should be treated the same by the current (unsound) alias analysis.

Reply:
Indeed, a fresh allocation certainly shouldn't alias more stuff than a `B*` parameter. I'll look at this issue later today.

Reply:
`git bisect`'ing revealed that 9cc3cda58eac4b4b4ad1fc7729a19ed182327faf actually was the commit that removed the allocation from all aliased memory.

Reply:
Merging in https://github.com/Semmle/ql/pull/3171 fixes the problem.